nucypher/datastore/datastore.py

```diff
@@ -14,269 +14,207 @@
-
-        deleted_records = 0
-        if result.count() > 0:
-            deleted_records = result.delete()
-        self.__commit(session=session)
-        return deleted_records
-
-    #
-    # Work Orders
-    #
-
-    def save_workorder(self,
-                       bob_verifying_key: UmbralPublicKey,
-                       bob_signature: Signature,
-                       arrangement_id: bytes,
-                       session=None
-                       ) -> Workorder:
-        """
-        Adds a Workorder to the keystore.
-        """
-        session = session or self._session_on_init_thread
-
-        # Get or Create Bob Verifying Key
-        fingerprint = fingerprint_from_key(bob_verifying_key)
-        key = session.query(Key).filter_by(fingerprint=fingerprint).first()
-        if not key:
-            key = self.add_key(key=bob_verifying_key)
-
-        new_workorder = Workorder(bob_verifying_key_id=key.id,
-                                  bob_signature=bob_signature,
-                                  arrangement_id=arrangement_id)
-
-        session.add(new_workorder)
-        self.__commit(session=session)
-        return new_workorder
-
-    def get_workorders(self,
-                       arrangement_id: bytes = None,
-                       bob_verifying_key: bytes = None,
-                       session=None
-                       ) -> List[Workorder]:
-        """
-        Returns a list of Workorders by HRAC.
-        """
-        session = session or self._session_on_init_thread
-        query = session.query(Workorder)
-
-        if not arrangement_id and not bob_verifying_key:
-            workorders = query.all()  # Return all records
+    pass
 
-        else:
 
-            # Return arrangement records
-            if arrangement_id:
-                workorders = query.filter_by(arrangement_id=arrangement_id)
+class DatastoreKey(NamedTuple):
+    """
+    Used for managing keys when querying the datastore.
+    """
+    record_type: Optional[str] = None
+    record_field: Optional[str] = None
+    record_id: Optional[Union[bytes, int]] = None
 
-            # Return records for Bob
-            else:
-                fingerprint = fingerprint_from_key(bob_verifying_key)
-                key = session.query(Key).filter_by(fingerprint=fingerprint).first()
-                workorders = query.filter_by(bob_verifying_key_id=key.id)
+    @classmethod
+    def from_bytestring(cls, key_bytestring: bytes) -> 'DatastoreKey':
+        key_parts = key_bytestring.decode().split(':')
+        with suppress(ValueError):
+            # If the ID can be an int, we convert it
+            key_parts[-1] = int(key_parts[-1])
+        return cls(*key_parts)
 
-            if not workorders:
-                raise NotFound
+    def compare_key(self, key_bytestring: bytes) -> bool:
+        other_key = DatastoreKey.from_bytestring(key_bytestring)
+        return self.record_type == (other_key.record_type or self.record_type) and \
```

Comment:
What happens if I have a `DatastoreKey`with actual values, and run `compare_key` on a `key_bytestring` and `cls` that returns `None` values for `record_type`, `record_field`, and `record_id`? Because of the `or self.<member>` in the `self.<member> == (other.<member> or self.<member>` comparison statements, this could still return `True` no?

Reply:
If I'm understanding you correctly, in practice, the `Datastore` cannot return a key with `None` types in those fields, so this case is not possible.

This is used for quickly iterating over the keys in lmdb for performing queries since the Python library can't do queries based on prefixes, so this is a relatively efficient implementation of that.

If you query for a partial key (such as `PolicyArrangement`), you're either going to find a key that matches it (meaning you have `PolicyArrangement` records) or you won't find a key that matches (meaning you have _no_ `PolicyArrangement` records). In the latter case, lmdb returns `None` and we always check, when setting the iterator to a lmdb key, that it _does not_ return `None`. If it does, we break our query loop.

So again, if I'm understanding you correctly, that's why we can effectively ignore that case. We never encounter keys that have a `None` and we effectively know that we will always have, at least, a `record_type`.

Reply:
Oh, I guess it would help to know that lmdb organizes keys lexicographically. Because of this, if I query for a key of `a` when I only have a key of `aardvark`, then my iterator will be placed at `aardvark`. Accordingly, if I query for a `PolicyArrangement:foobar_id`, but I only have a `PolicyArrangement:not_foobar_id`, then the cursor will return `None`. Hopefully that answers your question.

Reply:
It makes a bit more sense now why you do the `or self.<member>` check - it's to place the iterator at a relevant spot.

So is the reason you return `True` when keys aren't actually equal, is to have a pointer to a particular spot when so that you can potentially do a subsequent insertion of a new record? However, I guess this means whenever just querying for a record (not caring about insertion if the data does not exist) you have to ensure that the returned data is indeed what you wanted and not just the iterator pointing to a potential insertion point.

Have I understood that correctly?

Reply:
So the key is irrelevant to insertion since lmdb handles all that logic for us. However, lmdb does _not_ handle query logic, so that's predominantly what this class is for.

When performing a query, we need to match keys that fit a certain structure and that comes with a sense of precedence. Consider the following query keys (as `<RecordType>:<RecordField>:<RecordID>`):

1. `TestRecord`
2. `TestRecord:test_field`
3. `TestRecord:test_field:1234`

These keys are enumerated in descending order based on specificity; 1 is least specific and 3 is most specific.
We need to be able to match keys that get returned whilst iterating over the key-value store and know if they are _within_ a query or not. The 1st query key will return _all_ `TestRecord`s. The 2nd query key will return `TestRecord`s that have a `test_field` field. The 3rd query key will return a `TestRecord` that has a `test_field` and an id of `1234`.

To perform queries like the first and the second, we need some precedence in the query key structure. Wherever the "key slot" (meaning the relevant attr in `DatastoreKey`) is _not_ set, the key slot will be `None` by default. This acts as a precedence wildcard.

In Python, the logic `primary or secondary` will always return `primary` when it's `True`, otherwise it will return `secondary`. The `(other.<member> or self.<member>)` logic acts as a method of setting precedence in the key structure; we _always_ compare to (and prefer) the _given key_ if stated, otherwise we'll take the precedence of the _comparing key_ (the key calling the `compare_key` method).

In the `Datastore` query logic, our _given key_ is the query key (like those above) and the _comparing key_ is the key retrieved from lmdb.

Reply:
I added some comments for clarification in the `compare_key` method.

It's best to think of this process as _comparing specific keys to partial keys_ where we need to know if a specific key matches a given partial key.

Reply:
👍